core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java

```diff
@@ -1,86 +1,148 @@
 package fr.sncf.osrd.stdcm.preprocessing.implementation;
 
+import static fr.sncf.osrd.sim_infra.api.RawSignalingInfraKt.getRouteLength;
+import static fr.sncf.osrd.utils.KtToJavaConverter.toIntList;
+
 import com.google.common.collect.HashMultimap;
 import com.google.common.collect.Multimap;
 import fr.sncf.osrd.api.stdcm.STDCMRequest;
-import fr.sncf.osrd.infra.api.signaling.SignalingInfra;
-import fr.sncf.osrd.infra.api.signaling.SignalingRoute;
-import fr.sncf.osrd.stdcm.OccupancyBlock;
-import fr.sncf.osrd.train.RollingStock;
+import fr.sncf.osrd.envelope_sim.PhysicsRollingStock;
+import fr.sncf.osrd.sim_infra.api.BlockInfra;
+import fr.sncf.osrd.sim_infra.api.InterlockingInfraKt;
+import fr.sncf.osrd.sim_infra.api.RawSignalingInfra;
+import fr.sncf.osrd.sim_infra.api.Route;
+import fr.sncf.osrd.sim_infra.utils.BlockRecoveryKt;
+import fr.sncf.osrd.stdcm.OccupancySegment;
+import fr.sncf.osrd.utils.indexing.MutableStaticIdxArrayList;
+import fr.sncf.osrd.utils.units.Distance;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
 
 public class UnavailableSpaceBuilder {
+    private static final long SIGHT_DISTANCE = Distance.fromMeters(400);
 
-    /** Only used when we have missing signal data */
-    private static final double DEFAULT_SIGHT_DISTANCE = 400;
-
-    /** Computes the unavailable space for each route, i.e.
+    /** Computes the unavailable space for each block, i.e.
      * the times and positions where the *head* of the train cannot be.
-     * This considers existing occupancy blocks, the length of the train,
-     * and the routes that must be left available behind the train
-     *
+     * This considers existing occupancy segments, the length of the train,
+     * and the blocks that must be left available behind the train
      * <br/>
      * This is the first step to compute STDCM, the goal is to get rid of railway rules and extra complexity
-     * as soon as possible. After this step we can look for a single curve that avoids unavailable blocks. */
-    public static Multimap<SignalingRoute, OccupancyBlock> computeUnavailableSpace(
-            SignalingInfra infra,
+     * as soon as possible. After this step we can look for a single curve that avoids unavailable segment. */
+    public static Multimap<Integer, OccupancySegment> computeUnavailableSpace(
+            RawSignalingInfra infra,
+            BlockInfra blockInfra,
             Collection<STDCMRequest.RouteOccupancy> occupancies,
-            RollingStock rollingStock,
+            PhysicsRollingStock rollingStock,
             double marginToAddBeforeEachBlock,
             double marginToAddAfterEachBlock
     ) {
 
-        Multimap<SignalingRoute, OccupancyBlock> res = HashMultimap.create();
+        Multimap<Integer, OccupancySegment> res = HashMultimap.create();
```

Comment:
Not a big fan of naming the object to return `res` or `result`. In small methods, I'd say it's fine, but this is a big chunk.. Maybe replace it with `unavailableSpace` ?